### ISSUE TITLE
preloading 'En-US' model in a docker build?

### ISSUE THREAD

**Issue by buttchurch** [jxmorris12/language_tool_python#30]
Hi,

First I want to say thanks for creating this great tool. It's awesome!

I have an issue using the tool in a docker container on google cloud. I need to minimise runtime as much as possible, so I'd like to preload the model I need in the build phase rather than downloading it at the point of calling `tool = language_tool_python.LanguageTool('en-US')`. That way the program won't be waiting for the model to download.

I tried locating the download link in the GCP logs, and manually doing that step in the dockerfile like this:
```
ENV VERSION 5.0
RUN apt-get install -y wget && \
  apt-get update && apt-get install -y unzip && \
  wget https://www.languagetool.org/download/LanguageTool-$VERSION.zip && \
  unzip LanguageTool-$VERSION.zip && \
  rm LanguageTool-$VERSION.zip
```
The file downloaded, but when I called the tool command it started downloading again.

I'm no docker expert, but if there is a way to pre-download the model along with the tool I'd love to know how. Thanks!

**Comment by jxmorris12** [comment]
Hi @buttchurch -- that seems like a good strategy for pre-loading the languagetool binary. I think the missing step is that by default, language_tool_python unzips to the path `~/.cache/language_tool_python`. You can specify a new path by setting the `LTP_PATH` environment variable. So try

```bash
ENV VERSION 5.0
RUN apt-get install -y wget && \
  apt-get update && apt-get install -y unzip && \
  wget https://www.languagetool.org/download/LanguageTool-$VERSION.zip && \
  unzip LanguageTool-$VERSION.zip -d ~/.cache/language_tool_python && \
  rm LanguageTool-$VERSION.zip
```

and let me know if that works (not positive I invoked `unzip` correctly, but that's the right idea)

**Comment by buttchurch** [comment]
Thanks so much – this worked!

**Comment by buttchurch** [comment]
Hi, I'm sorry to reopen this issue but I've just realised that the code fix didn't work – I just hadn't turned language_tool_python back on in my code, so it wasn't being called at all. I've got it working again, but calling `language_tool_python.LanguageTool('en-US')` is still prompting me to re-download the model in full.

Any other ideas for how to pre-load the model? Again, thanks for your helpful support.

**Comment by jxmorris12** [comment]
No problem @buttchurch I can help!

I'm kind of convinced still that the file just isn't pre-downloaded to the correct place. Can you try `ls ~/.cache/language_tool_python` and let me know what the output is? If LanguageTool is downloaded to that folder, I'll have to think some more.

**Comment by buttchurch** [comment]
Hi Jack, thanks so much for your help. I put in an echo statement to the dockerfile to print that output, and:

`~/.cache/language_tool_python` contains only the `LanguageTool-5.0` directory.

But i watched the zip file being unpacked, and it is expanding in that 5.0 directory, to `.cache/language_tool_python/LanguageTool-5.0`. It mostly consists of files in the `org` and `libs` directories.

Does this give you a clue of what I'm doing wrong?

Thanks!
Britton

**Comment by jxmorris12** [comment]
Awesome, thanks, this actually might be enough information to solve the issue! @buttchurch can you try changing the 5.0 to 5.2? 

5.2 is the latest version, so right now LTP probably checks for 5.2, only sees 5.0, and automatically downloads 5.2. That's my hypothesis, anyway. Let me know if it works!!

**Comment by buttchurch** [comment]
Hi Jack! I'm sad to say this didn't work. I tried simply changing it to `ENV=5.2` in the snippet above, which didn't work.

I also tried reverting to `ENV=5.0` while locking the requirements.txt entry to `language_tool_python==2.5.0`, which didn't work either.

Any other thoughts as to what might be going on? Anything else I can try for you? Perhaps I should try `ENV=2.5.2` instead in the Dockerfile?

Thanks very much,
Britton

**Comment by jxmorris12** [comment]
Sorry Britton, this must be frustrating.

This is what my cache folder looks like:

<img width="550" alt="Screen Shot 2021-01-12 at 10 06 31 AM" src="https://user-images.githubusercontent.com/13238952/104332133-d8ee9a80-54bd-11eb-8514-d30ab005b1b0.png">

Versions 5.0 and 5.1 are older and now outdated. All you need is LanguageTool-5.2. If you can download and unzip it to that path on your docker image, the script shouldn't have to download languagetool again.

**Comment by buttchurch** [comment]
Hey Jack! I really appreciate you sticking with me and being so responsive.

My setup looked exactly like you said it should, so I finally tried simply calling the Python command `language_tool_python.LanguageTool('en-US')` in the Dockerfile to see if it would somehow setup differently. It prompted the model download, so I could feel 100% confident that it was downloaded and configured correctly in the build.

...then when I went to run my script in the Docker container, the script _still_ prompted a download. So I can only conclude that your initial instructions were correct, and the problem is something with how I'm using Docker.

Thanks for your help, and sorry for all the run-around. I'll close this issue now.

Cheers!
B
